### PR TITLE
Issue #2391 - JSON string encoding fix

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java [hunk 1]
```diff
@@ -93,12 +92,23 @@
     public final static JSON DEFAULT = new JSON();
 
     private Map<String, Convertor> _convertors = new ConcurrentHashMap<String, Convertor>();
+    private StringEncoder _stringEncoder = new JSONDefaultStringEncoder();
```
Comment (gregw) [inline jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java]: This should be final and initialized with constructor injection

File: jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java [hunk 2]
```diff
@@ -131,6 +141,21 @@ public static void registerConvertor(Class forClass, Convertor convertor)
         DEFAULT.addConvertor(forClass,convertor);
     }
 
+    /**
+     * Set the preferred String Encoder for JSON
+     *
+     * @param encoder the encoder implementation
+     */
+    public static void setStringEncoder(StringEncoder encoder)
```
Comment (gregw) [inline jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java]: it is asymmetric to have a plugable encoder but not a plugable decoder

File: jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSONEscapedStringEncoder.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util.ajax;
+
+import java.io.IOException;
+
+@SuppressWarnings("Duplicates")
+public class JSONEscapedStringEncoder implements JSON.StringEncoder
```
Comment (gregw) [inline jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSONEscapedStringEncoder.java]: It's not really an encoder but a quoter.

File: jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java (hunk 1 above)
Comment (joakime) [reply jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java]: Can't do it effectively through the constructor, as JSON is essentially a singleton. :-(

**Comment by joakime** [comment]
continuing conversation in Issue #2391

**Comment by joakime** [comment]
One thing I'm not 100% happy with is the escaping of the solidus U+002F
I think that is also optional, but the RFC makes it unclear.

File: jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java (hunk 1 above)
Comment (gregw) [reply jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/JSON.java]: Then  a setter is even worse, as that can change encoding for other threads using the singleton.   Perhaps it needs to be an optional parameter to the methods on JSON?

**Pull request merged**